artemis-core-client/src/main/java/org/apache/activemq/artemis/core/protocol/core/impl/ResponseCache.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.activemq.artemis.core.protocol.core.impl;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.apache.activemq.artemis.api.core.ActiveMQInterruptedException;
+import org.apache.activemq.artemis.core.protocol.core.Packet;
+import org.apache.activemq.artemis.core.protocol.core.ResponseHandler;
+
+public class ResponseCache {
+
+   private final AtomicInteger writerPointer = new AtomicInteger(0);
+   private final AtomicInteger sequence = new AtomicInteger(0);
+
+   private final Packet[] store;
+   private ResponseHandler responseHandler;
+
+   public ResponseCache(int size) {
+      this.store = new Packet[size];
+   }
+
+   public long add(Packet packet) {
+      int pointer = writerPointer.getAndUpdate(operand -> {
+         Packet p = store[operand];
+         if (p != null) {
+            return operand;
+         }
+         packet.setCorrelationID(correlationID(operand, sequence.incrementAndGet()));
+         store[operand] = packet;
+         return operand + 1 == store.length ? 0 : operand + 1;
+      });
+
+      if (pointer(packet.getCorrelationID()) != pointer) {
+         throw new ActiveMQInterruptedException("unable to send due to buffer full");
+      }
+
+      return packet.getCorrelationID();
+   }
+
+   public void handleResponse(Packet response) {
+      long correlationID = response.getCorrelationID();
+      int pointer = pointer(correlationID);
+      if (pointer > -1 && pointer < store.length) {
+         Packet p = store[pointer];
+         if (p != null && p.getCorrelationID() == correlationID) {
+            store[pointer] = null;
+         }
+         if (responseHandler != null) {
+            responseHandler.responseHandler(p, response);
+         }
+      }
+   }
+
+   public long correlationID(int pointer, int sequence) {
```

Comment:
I would make this method static and uses `(pointer & 0xFFFF_FFFFL) << 32` to make explicit what is happing during the upcast